File: vectorization/matvector/fortran/solutions/elemental_best/multiply.f90
!* ========================================================================== *
!*
!* SAMPLE SOURCE CODE - SUBJECT TO THE TERMS OF SAMPLE CODE LICENSE AGREEMENT,
!* http://software.intel.com/en-us/articles/
!*        intel-sample-source-code-license-agreement/
!*
!* Copyright 2010-2017 Intel Corporation
!*
!* THIS FILE IS PROVIDED "AS IS" WITH NO WARRANTIES, EXPRESS OR IMPLIED,
!* INCLUDING BUT NOT LIMITED TO ANY IMPLIED WARRANTY OF MERCHANTABILITY,
!* FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT OF INTELLECTUAL
!* PROPERTY RIGHTS.
!*
!* ========================================================================== *
#ifdef __MIC__
!DIR$ message "You are compiling " __FILE__ " for the coprocessor now!"
#endif

subroutine matvec(rows, cols, a, b, x)
    use Global ! inc_i & inc_j
    implicit none
    integer rows, cols, i, j
    real*8, contiguous :: a(:,:), b(:), x(:)
!DIR$ ASSUME_ALIGNED a:64, b:64, x:64
!DIR$ ASSUME (MOD(rows, 64) .EQ. 0)

    do i = 1, cols, inc_i
!DIR$ VECTOR ALIGNED
        b(i) = sum(mult(a(1:rows:inc_j, i), x(1:rows:inc_j)))
    enddo

    contains
        elemental real*8 function mult(a, x)
            implicit none
            real*8, intent(in) :: a, x
            mult = a * x
        end function
end subroutine

File: vectorization/matvector/c/solutions/restrict/multiply.c
/* ========================================================================== *
!*
!* SAMPLE SOURCE CODE - SUBJECT TO THE TERMS OF SAMPLE CODE LICENSE AGREEMENT,
!* http://software.intel.com/en-us/articles/intel-sample-source-code-license-agreement/
!*
!* Copyright 2010-2018 Intel Corporation
!*
!* THIS FILE IS PROVIDED "AS IS" WITH NO WARRANTIES, EXPRESS OR IMPLIED,
!* INCLUDING BUT NOT LIMITED TO ANY IMPLIED WARRANTY OF MERCHANTABILITY,
!* FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT OF INTELLECTUAL
!* PROPERTY RIGHTS.
!*
!* ========================================================================== */

#include "multiply.h"

void matvec(unsigned int rows, unsigned int cols,
            FTYPE (*a)[cols], FTYPE * restrict b, FTYPE *x, FTYPE *c)
{
    int i, j;

    // ATTENTION:
    // This is just an example to demonstrate the effects in case the compiler
    // does not have information at compile time. As this is a matrix vector
    // multiplication, there is no reason to use non-unit strides here.
    // For other algorithms: Think about whether it is possible to reduce
    // runtime variables to compile time constants (e.g. with templates,
    // dedicated function implementations, reduction of parameters, etc.).
    enum { inc_i = 1, inc_j = 1 };

    for (i = 0; i < rows; i += inc_i) {
        for (j = 0; j < cols; j += inc_j) {
            b[i] += a[i][j] * x[j] + c[i];
        }
    }
}
